trainee_template/hierarchy.amlg

```diff
@@ -0,0 +1,368 @@
+(null
+
+	;Helper method to create the actual full name path by interleaving ".trainee_container" with each name in the path
+	#ConvertChildPathToActualPath
+	(weave (range traineeContainer 1 (size path) 1) path)
+
+
+	;return full path to specified subtrainee in the format of (list ".trainee_container" child1 ".trainee_container" child2)
+	;parameters:
+	; name_path: optional,  list of strings specifying the user-friendly path to the child subtrainee.
+	;		May be an incomplete path as long as it uniquely identifies the trainee.
+	; id: optional, unique id of the subtrainee. Ignored if namp_path is specified.
+	#ResolveChildPath
+	(seq
+		;use id instead of name_path if provided only id
+		(if (and (= 0 (size name_path)) id)
+			(conclude
+				(if (contains_index containedTraineeIdToNamePathMap id)
+					(call ConvertChildPathToActualPath (assoc
+						path (get containedTraineeIdToNamePathMap id)
+					))
+
+					"Invalid trainee id specified."
+				)
+			)
+		)
+
+		(declare (assoc name (apply "concat" name_path) ))
+
+		;keep all those paths that match the provided name
+		(declare (assoc
+			matching_paths
+				(filter
+					(lambda (= name (substr (current_index) name)) )
+					containedTraineeNamePathStringToListMap
+				)
+		))
+
+		(if (= 1 (size matching_paths))
+			(conclude
+				(call ConvertChildPathToActualPath (assoc
+					path (first (values matching_paths))
+				))
+			)
+
+			(= 0 (size matching_paths))
+			(conclude "Invalid name path specified.")
+		)
+
+		;if there is more than one matching path, either name_path refers to a non-leaf node
+		;or it's too ambiguous.  If it's a non-leaf node, one of the matching_paths will have it as the last name listed in the path
+		(assign (assoc
+			matching_paths
+				(filter
+					(lambda (= (last name_path) (last (current_value))))
+					matching_paths
+				)
+		))
+
+		(if (= 1 (size matching_paths))
+			(call ConvertChildPathToActualPath (assoc
+				path (first (values matching_paths))
+			))
+
+			"Ambiguous name path specified, more than one trainee matches the specified name path."
+		)
+	)
+
+	;Return the flag whether the subtrainee at the provided full path is stored inside (true) or independently (false)
+	#IsChildInternal
+	(get
+		childTraineeIsContainedMap
+		(apply "concat" (filter (lambda (!= traineeContainer (current_value))) child_trainee_path))
+	)
+
+	;Method to update this trainee's Parent's is_contained status
+	;parameters:
+	; child_trainee_path: path to child trainee
+	; is_contained: boolean, true means this child is contained inside traineeContainer, false means it is stored and loaded independently
+	#UpdateParentsIsContainedMap
+	(let
```

Comment:
Should all this be wrapped in some logic depending on is_contained? all this logic only seems to apply if it is contained, but maybe I'm lost

Reply:
as the method description says, it updates the trainee's parent's 'is_contained' status to false since this trainee is being exported/saved as an independent. i'll make the method comment more clear